src/h.js

```diff
@@ -1,22 +1,29 @@
-export function h(tag, data) {
-  var node
-  var stack = []
-  var children = []
+function addChild(children, val) {
+  if (val == null || typeof val === 'boolean') return
 
-  for (var i = arguments.length; i-- > 2; ) {
-    stack[stack.length] = arguments[i]
+  if (Array.isArray(val)) {
+    for (var j = 0; j < val.length; j++) {
+      children.push(val[j])
+    }
   }
+  else {
+    children.push(typeof val === 'number' ? val + '' : val)
+  }
+}
 
-  while (stack.length) {
-    if (Array.isArray((node = stack.pop()))) {
-      for (var i = node.length; i--; ) {
-        stack[stack.length] = node[i]
+export function h(tag, data, values) {
+  var children = []
+
+  if (arguments.length > 2) {
+    if (Array.isArray(values)) {
```

Comment:
Factoring this piece is really tempting I admit 👀
But `iter = isArr ? values : arguments` is a honeypot. `arguments` is very different from a JS array and does not use the same stubs to iterate over. So if `iter` is either `arguments` or an array, the function is likely to never be optimized. Here is a comparison of the 2 versions: http://quick.as/qnGBfdboR

> This is because for-loops are always (currently) faster if i and len are declared early.

AFAIK code-motion takes this in charge: `i` is hoisted out of the loop, and `len` is probably too because the inner code is invariant and does not mutate the array, so the compiler can safely make the assumption `length` will never change.